### PR TITLE
feat: vSphere cloud-init / build command

### PR DESCRIPTION
**Pull request by fatz**
**What problem does this PR solve?**:
https://d2iq.atlassian.net/browse/D2IQ-93355

Support different setups by issuing cloud-init:
- support only public key given and expect ssh-agent heaving private key
- support neither public or private key given. Generate a key that will be used for authentication
- any username can be specified

Add the missing vsphere build command to have the same simple interface as with other clouds

```
bin/konvoy-image build vsphere --help
build and provision vsphere images

Usage:
  konvoy-image build vsphere <image.yaml> [flags]

Examples:
vsphere --datacenter dc1 --cluster zone1 --datastore nfs-store1 --network public --template=d2iq-base-templates/d2iq-base-CentOS-7.9 images/ami/centos-79.yaml

Flags:
      --cluster string               vSphere cluster to be used. Alternatively set host (REQUIRED)
      --containerd-version string    the version of containerd to install
      --datacenter string            vSphere datacenter (REQUIRED)
      --datastore string             datastore used to build and store the KIB image (REQUIRED)
      --dry-run                      do not create artifacts, or delete them after creating. Recommended for tests.
      --extra-vars strings           flag passed Ansible's extra-vars
      --folder string                deploy KIB image into this folder
  -h, --help                         help for vsphere
      --host string                  vSphere host to be used. Alternatively set cluster (REQUIRED)
      --kubernetes-version string    The version of kubernetes to install. Example: 1.21.6
      --network string               vSphere network to start KIB build; KIB expects DHCP on this network; Ensure the machine running KIB has access to this network (REQUIRED)
      --overrides strings            a comma separated list of override YAML files
      --packer-manifest string       provide the path to a custom packer manifest
      --packer-on-error string       [advanced] set error strategy for packer. strategies [cleanup, abort, run-cleanup-provisioner]
      --packer-path string           the location of the packer binary (default "packer")
      --resource-pool string         vSphere resource pool to be used to build KIB image
      --ssh-privatekey-file string   path to ssh private key which will be used to loginto the template
      --ssh-publickey string         SSH public key which will be deployed using cloud-init. Ensure to set ssh-privatekey-file or load the private key into ssh-agent
      --ssh-username string          Username to be used with the vSphere template
      --template string              Base template to be used. Can include folder. <templatename> or <folder>/<templatename> (REQUIRED)
      --work-dir string              path to custom work directory generated by the generate command

Global Flags:
      --color     enable color output (default true)
  -v, --v int     select verbosity level, should be between 0 and 6
      --verbose   enable debug level logging (same as --v 5)
```

**Which issue(s) does this PR fix?**:

* https://jira.d2iq.com/browse/D2IQ-NUMBER

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: cmd/konvoy-image/cmd/vsphere.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		&vsphereArgs.Host,
+		"host",
+		"",
+		"vSphere host to be used. Alternatively set cluster (REQUIRED)",
+	)
+	fs.StringVar(
+		&vsphereArgs.Datastore,
+		"datastore",
+		"",
+		"datastore used to build and store the KIB image (REQUIRED)",
```
Comment (supershal) [inline cmd/konvoy-image/cmd/vsphere.go:97]:
```suggestion
		"datastore used to build and store the vSphere image template (REQUIRED)",
```

File: cmd/konvoy-image/cmd/vsphere.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	)
+	fs.StringVar(
+		&vsphereArgs.Network,
+		"network",
+		"",
+		"vSphere network to start KIB build; KIB expects DHCP on this network; "+
+			"Ensure the machine running KIB has access to this network (REQUIRED)",
```
Comment (supershal) [inline cmd/konvoy-image/cmd/vsphere.go:104]:
```suggestion
		"vSphere network used to build image template. "+
			"Ensure the host running the command has access to this network (REQUIRED)",
```

Comment (supershal) [reply cmd/konvoy-image/cmd/vsphere.go:104]: Not sure if we should mention about DHCP. may be our documentation can cover that?

File: cmd/konvoy-image/cmd/vsphere.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+		"resource-pool",
+		"",
+		"vSphere resource pool to be used to build KIB image",
+	)
+
+	fs.StringVar(
+		&vsphereArgs.ResourcePool,
+		"ssh-privatekey-file",
+		"",
+		"path to ssh private key which will be used to loginto the template",
```
Comment (supershal) [inline cmd/konvoy-image/cmd/vsphere.go:123]:
```suggestion
		"path to ssh private key which will be used to log into the template",
```

File: cmd/konvoy-image/cmd/vsphere.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+		"ssh-publickey",
+		"",
+		"SSH public key which will be deployed using cloud-init. Ensure to set ssh-privatekey-file or load the private key into ssh-agent",
+	)
+
+	fs.StringVar(
+		&vsphereArgs.SSHUserName,
+		"ssh-username",
+		"",
+		"Username to be used with the vSphere template",
```
Comment (supershal) [inline cmd/konvoy-image/cmd/vsphere.go:137]:
```suggestion
		"username to be used with the vSphere image template",
```

File: cmd/konvoy-image/cmd/vsphere.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	gFlags.userArgs.VSphere = &app.VSphereArgs{}
+	addVSphereArgs(fs, gFlags.userArgs.VSphere)
+}
+
+func addVSphereArgs(fs *flag.FlagSet, vsphereArgs *app.VSphereArgs) {
+	fs.StringVar(
+		&vsphereArgs.Template,
+		"template",
+		"",
+		"Base template to be used. Can include folder. <templatename> or <folder>/<templatename> (REQUIRED)",
```
Comment (supershal) [inline cmd/konvoy-image/cmd/vsphere.go:71]:
```suggestion
		"base template to be used. Can include folder. <templatename> or <folder>/<templatename> (REQUIRED)",
```

File: pkg/app/config.go [hunk 6]
```diff
@@ -413,6 +413,12 @@ func MergeUserArgs(config Config, userArgs UserArgs) error {
 		}
 	}
 
+	if userArgs.VSphere != nil {
+		if err := MergeVSphereUserArgs(config, userArgs.VSphere); err != nil {
+			return fmt.Errorf("failed to set gcp args: %w", err)
```
Comment (supershal) [inline pkg/app/config.go:418]:
```suggestion
			return fmt.Errorf("failed to set vSphere args: %w", err)
```

File: pkg/packer/manifests/vsphere/packer.pkr.hcl [hunk 7]
```diff
@@ -353,8 +367,36 @@ locals {
...
+
+  # inject generated key if no agent auth or private key is given
+  ssh_private_key_file = var.ssh_private_key_file != "" ? var.ssh_private_key_file : local.ssh_agent_auth ? "" : data.sshkey.kibkey.private_key_path
+  # when ssh_private_key_file uses the generated key inject its public key
+  ssh_public_key = local.ssh_private_key_file == data.sshkey.kibkey.private_key_path ? data.sshkey.kibkey.public_key : var.ssh_public_key
+
+  # prepare cloud-init
+  cloud_init = <<EOF
+#cloud-config
+users:
```
Comment (supershal) [inline pkg/packer/manifests/vsphere/packer.pkr.hcl:382]: awesome

File: pkg/app/config.go (hunk 6 above)
Comment (fatz) [reply pkg/app/config.go:418]: thanks... obviously copy+paste error. good catch!

**Review by supershal — APPROVED** [review]
Thank you for going above and beyond to writing this feature enhancement.

File: Dockerfile [hunk 8]
```diff
@@ -23,6 +23,7 @@ RUN apk add --no-cache \
         py3-cryptography \
         py3-pip \
         py3-wheel \
+        xorriso \
```
Comment (supershal) [inline Dockerfile:26]: we will not need this package as we are not going to build template using ISO.

Comment (fatz) [reply Dockerfile:26]: well we do... thats how kickstart is being injected

Comment (supershal) [reply Dockerfile:26]: got it.

File: .github/workflows/aws-e2e.yaml [hunk 9]
```diff
@@ -24,7 +24,6 @@ jobs:
         - "flatcar"
         - "ubuntu 18.04"
         - "ubuntu 20.04"
-        - "rocky 9.1"
```
Comment (faiq) [inline .github/workflows/aws-e2e.yaml:27]: Why is this being removed?

File: .github/workflows/release-ami.yaml [hunk 10]
```diff
@@ -37,6 +37,8 @@ jobs:
             buildConfig: "offline-fips"
           - os: "rocky 9.1"
             buildConfig: "offline"
+          - os: "rocky 9.1"
+            buildConfig: "offline-fips"
```
Comment (faiq) [inline .github/workflows/release-ami.yaml:41]: i dont think we support this?

**Review by faiq — CHANGES_REQUESTED** [review]
there seem to be a lot of unrelated changes from a rebase or something

**Comment by fatz** [comment]
i'm closing this as  #800 is cloning this and got merged

**Pull request closed without merging**